### DISCUSSION TITLE
Cloudflare's Rocket Loader modifying the HTML And Breaking Href's

### DISCUSSION THREAD

**Discussion by samdotxml** [retypeapp/retype#380] [General]
Hello there.

I hope this is the right place to "report" this. I have recently setup a documentation site with retype. I then had following issue:
Upon clicking on some button to navigate to a different section in the docs, i would be redirected to the wrong page which does not exist.

After a long time of "debugging" i found out following issue:
If you use Cloudflare, the HTML that gets sent from the server to the client, gets altered by Cloudflare's Rocket Loader and hrefs can be modified.

I checked out the networks tab in my browser. The server sends the correct html but a javascript function (Rocket Loader) alters it and breaks the hrefs.

Here is a page i randomly found that proofs this. It's from another documentation software and it's on their FAQ.
https://www.helpndoc.com/documentation/html/TheHTMLhelpisbrokenwhenhostedbyC.html

So if anyone faces the same issue you have to disable the Rocket Loader in the Cloudflare settings. **I just want to say this clearly, this is not an issue from retype, it's cloudflare's / rocket loader's fault**

Maybe retype could add this to the FAQ or some common issues list.

Apologies if this has already been mentioned but before posting this i found 0 results in the issues or discussion threads.

**Comment by DannyBen** [comment]
Thanks - good to know.
I wonder if this is something that Cloudflare should be made aware of - or why it happens in the first place. There is nothing too complex in Retype that justifies breaking like this.

**Reply by samdotxml** [reply]
> Thanks - good to know. I wonder if this is something that Cloudflare should be made aware of - or why it happens in the first place. There is nothing too complex in Retype that justifies breaking like this.

I don't know what exactly is causing this. According to an article on Cloudflare's website, it "prioritizes" content and on the website. I don't think Cloudflare needs to be made aware of it, perhaps it's an unfortunate coincidence that the two don't work. 
I think it'd be great to add this to the FAQ or maybe the ignore tag will fix the issue, the article mentions how to use it.

Source: https://support.cloudflare.com/hc/en-us/articles/200168056-Understanding-Rocket-Loader

**Comment by geoffreymcgill** [comment]
For reference, another Cloudflare Rocket Loader issue was reported a while back...

https://github.com/retypeapp/retype/issues/117

The solution was also disabling Rocket Loader.

**Reply by samdotxml** [reply]
Didn't see this one. Thanks for sharing and apologies for double posting. 

Though it could be taken into consideration to use the exclude tags on the generated doc sites. When disabling the Cloudflare Rocket Loader, you disable it for the entire domain. Maybe other services reachable under a different sub-domain are in need or gain performance with the Rocket Loader.

**Reply by geoffreymcgill** [reply]
We are not entirely sure if adding the exclude tags will help with this issue. Rocket Loader appears to have some kind of conflict with the [Turbo](https://turbo.hotwired.dev/) library that is used to create a single-page-application experience within Retype. 

There is a new release of Turbo available, so we're going to upgrade Retype and run some tests using Cloudflare with Rocket Loader enabled. The retype.com domain uses Cloudflare, so we should theoretically be able to reproduce any Cloudflare related issues right on the retype.com website.

**Comment by geoffreymcgill** [comment]
As of the upcoming v3.1.0 release, Retype now supports Cloudflare Rocket Loader. There are no changes required in your project, other than building with Retype v3.1.0 once it is released. 

Even though Retype is now working well, we would still recommend disabling Cloudflare Rocket Loader within your Cloudflare admin panel.
